packages/riverpod/lib/src/framework/auto_dispose.dart

```diff
@@ -20,6 +21,33 @@ abstract class AutoDisposeProviderReference extends ProviderReference {
   T watch<T>(RootProvider<Object?, T> provider);
 }
 
+/// A [ProviderReference] for providers that are automatically destroyed when
+/// no-longer used.
+///
+/// The difference with [ProviderReference] is that it has an extra
+/// [maintainState] property, to help determine if the state can be destroyed
+///  or not.
+///  It will also allow the call [setState] to update the currently exposed state
+///  of the provider.
+abstract class AutoDisposeProviderReferenceAdvanced<Listened>
```

Comment:
I don't like "advanced" as a name. It doesn't mean much.

What about *EditableProviderReference?

Reply:
Or maybe better:
Name this `ProviderReference`, and rename the variants that don't have a `setState` into `ImmutableProviderReference`

Reply:
I don't like "advanced" either :D It was meant to be a temporary name :) But i _do_  like `ImmutableProviderReference` so i will go with this one.